envoy/event/dispatcher.h

```diff
@@ -228,8 +228,9 @@ class Dispatcher : public DispatcherBase, public ScopeTracker {
    * @return Network::DnsResolverSharedPtr that is owned by the caller.
    */
   virtual Network::DnsResolverSharedPtr
-  createDnsResolver(const std::vector<Network::Address::InstanceConstSharedPtr>& resolvers,
-                    const envoy::config::core::v3::DnsResolverOptions& dns_resolver_options) PURE;
+  createDnsResolver(
+      envoy::config::core::v3::DnsResolutionConfig& dns_resolution_config,
+      envoy::config::core::v3::TypedExtensionConfig& dns_resolver_config) PURE;
```

Comment:
The comments above should be updated.
I would also suggest renaming the variables, as `dns_resoultion_config` and `dns_resolver_config` imply the same thing. I think it might make sense to rename the second to be something more specific to typed extension (maybe `dns_resolution_typed_config`).

Reply:
I would take it a step further and create a helper method (i.e. in `source/network/dns_resolvers/common/`) to handle backward compatibility and produce the `envoy::config::core::v3::TypedExtensionConfig`. The this interface would be cleaner with just a single parameter.

Reply:
Thanks for the comments!

Do you mean the logic here should be: 

If typed_dns_resolver_config is in place, just pass this parameter down to the dispatcher:createDnsResolver() function. 
If it is missing,  then check the run-time config: use_apple_api_for_dns_lookups. If it is enabled, then pack an empty apple_dns_resolver message into typed_dns_resolver_config. If  use_apple_api_for_dns_lookups is not enabled, then pack the dns_resolution_config  into typed_dns_resolver_config. In both case, we just pass  one parameter: typed_dns_resolver_config down to the dispatcher:createDnsResolver()  function.   And inside dispatcher:createDnsResolver(), we don't have to check the typed_dns_resolver_config emptiness, just call the factory functions.

Is this understanding correct? 


If it is, Interface wise, this is simpler.  The draw back is that it needs to have the " if  (use_apple_api_for_dns_lookups)  PackFrom()... "logic to all the callers , instead of having it in a central place . 

Please comment.

Reply:
I meant moving the logic of dealing with the `envoy::config::core::v3::DnsResolutionConfig`, Apple DNS and `envoy::config::core::v3::TypedExtensionConfig` into a separate helper function. After that the createDnsResolver will have only one parameter `envoy::config::core::v3::TypedExtensionConfig`.

Reply:
Good idea, added a template function to construct the typed_dns_resolver_config based on different configurations (bootstrap, cluster, dns_cache_config, dns_filter_config).